File: NumberOfIslands_DFS.java
// Time Complexity : O(m * n) --> where m and n are lengths of input 2D matrix
// Space Complexity : O (m * n)
// Did this code successfully run on Leetcode (200): Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    int dirs[][];
    int m;
    int n;
    public int numIslands(char[][] grid) {
        // edge case
        if (grid == null || grid.length == 0) return 0;
        
        m = grid.length;
        n = grid[0].length;
        dirs = new int[][] {{0,1}, {0,-1}, {1,0}, {-1,0}};
        int count = 0;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }
    
    private void dfs(char[][] grid, int i, int j) {
        // base case
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == '0') return;
        
        // logic
        grid[i][j] = '0';        
        for (int dir[] : dirs) {
            int r = i + dir[0];
            int c = j + dir[1];
            dfs(grid, r, c);
        }
    }
}

File: numberOfIslandsDFS.py
class Solution:
    def __init__(self):
        self.dirs = None
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return None
        
        self.dirs = ((0,1),(1,0),(0,-1),(-1,0))
        count = 0
        m = len(grid)
        n = len(grid[0])
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    count+=1
                    self.dfs(grid, i, j, m, n)
                
        
        return count
    
    def dfs(self, grid, i, j, m, n):
        #base
        if i<0 or i==m or j<0 or j==n or grid[i][j] != '1':
            return

        grid[i][j] = '0'
        #logic
        for dir in self.dirs:
            r = dir[0]+i
            c = dir[1]+j
            self.dfs(grid, r,c,m,n)
            
# time complexity is o(n) and space complexity is min(r,c)
        
